_posts/2021-02-10-native-k8s-with-ha.md

```diff
@@ -0,0 +1,112 @@
+From Flink 1.12, we leverage these features to make running a HA-configured Flink cluster on Kubernetes more convenient to users.
+
+<center>
+<img vspace="8" style="width:75%" src="{{site.baseurl}}/img/blog/2021-02-10-native-k8s-with-ha/native-k8s-ha-architecture.png" />
+</center>
+
+The above diagram shows the architecture of Flink’s Kubernetes HA service, which works as follows:
+
+ 1. For the leader election, a set of eligible JobManagers is identified. They all race to declare themselves as the leader, with one winning and becoming the active leader. The active JobManager then continually "heartbeats" to renew its position as the leader. In the meantime, all other standby JobManagers periodically make new attempts to become the leader — this ensures that the JobManager could failover quickly. Different components (e.g. ResourceManager, JobManager, Dispatcher, RestEndpoint) have separate leader election services and ConfigMaps.
+ 2. The active leader publishes its address to the ConfigMap. It’s important to  note that Flink uses the same ConfigMap for contending lock and storing the leader address. This ensures that there is no unexpected change snuck in during a periodic update.
+ 3. The leader retrieval service is used to find the active leader’s address and allow the components to then register themselves. For example, TaskManagers retrieve the address of ResourceManager and JobManager for the registration and offering slots. Flink uses a Kubernetes watch in the leader retrieval service — once the content of ConfigMap changes, it usually means that the leader has changed, and so the listener can get the latest leader address immediately.
+ 4. All other meta information (e.g. running jobs, job graphs, completed checkpoints and checkpointer counter) will be directly stored in the corresponding ConfigMaps. Only the leader can update the ConfigMap. The HA data will only be cleaned up once the Flink cluster reaches the global terminal state. Please note that only the pointers are stored in the ConfigMap; the concrete data has been stored in DistributedStorage. This level of indirection is necessary to keep the amount of data in ConfigMap small (ConfigMap is built for data less than 1MB whereas state can grow to multiple GBs).
+
+# Example: Application cluster with HA
+
+Please note that you need a running Kubernetes cluster and to get kube config properly set to follow along.
+You can use `kubectl get nodes` for verifying that you’re all set.
+In this blog post, we’re using [minikube](https://minikube.sigs.k8s.io/docs/start/) for local testing.
+
+1. Build a Docker image with the Flink job (my-flink-job.jar) baked in
+
+        FROM flink:1.12.1
+        RUN mkdir -p $FLINK_HOME/usrlib
+        COPY /path/of/my-flink-job.jar $FLINK_HOME/usrlib/my-flink-job.jar
+
+    Use the above Dockerfile to build a user image and then push it to your remote image repository.
+
```

Comment:
```suggestion
docker built -t <user-image> .
docker push <user-image>
```
Should we add the docker commands here as well? This way, we would have a step by step list of commands that need to be called by the user. This way, we also make clear what `<user-image>` refers to in the next command.